src/index.ts

```diff
@@ -1458,4 +1458,63 @@ export function VerifyWitness(
     }
   }
   throw new VerificationError(`Unexpected end of proof`);
-}
\ No newline at end of file
+}
+
+/**
+ * Verifies a witness against a staleRoot, constructs newest witness from the recentState and verifies it
+ * @param staleRoot   stale root
+ * @param key         key being read
+ * @param witness     witness against a stale root
+ * @param recentState Merkle Paticia Tree horizontally caches nodes to some depth and veritically caches most recent keys
+ */
+export function VerifyStaleWitness(
+    staleRoot: Buffer, key: Buffer, witness: Witness<Buffer>, recentState: MerklePatriciaTree) {
+  // Verify if the witness is valid against the stale root
+  VerifyWitness(staleRoot, key, witness);
+  // If the stale root and recent root match, return the witness
+  if (staleRoot.compare(recentState.root) === 0) {
+    return witness;
+  }
+  // Check if nodes match at any feasible depth (search happens till available cacheDepth)
+  const recentWitness: Witness<Buffer> = {proof: [], value: null};
+  const result: SearchResult = recentState.search(key);
+  // If the key was vertically cached in the recentState; search will find the path along the key
+  if (result.remainder.length === 0 && result.node !== null) {
+    recentWitness.value = result.node.value;
+    const witProof: Buffer[] = [];
+    for (const [idx, witNode] of result.stack.entries()) {
+      const rlp = witNode.getRlpNodeEncoding(recentState.convertValue);
+      if (rlp.length >= 32 || (idx === 0)) {
+        witProof.push(rlp);
+      }
+    }
+    recentWitness.proof = witProof;
+    VerifyWitness(recentState.root, key, recentWitness);
+    return;
+  }
+  // If it was not vertically cached, then the key wasn't changed so the node hashes should match at some depth
+  recentWitness.proof = [];
+  recentWitness.value = witness.value;
+  const oldNodeHashes = [];
+  for (const serializedOldNode of witness.proof) {
+    oldNodeHashes.push(hashAsBuffer(HashType.KECCAK256, serializedOldNode));
+  }
+  for (const [idx, witNode] of result.stack.entries()) {
+    const rlp = witNode.getRlpNodeEncoding(recentState.convertValue);
+    let recentHash = Buffer.from("");
+    if (rlp.length >= 32 || (idx === 0)) {
+      recentWitness.proof.push(rlp);
+      recentHash = hashAsBuffer(HashType.KECCAK256, rlp);
```

Comment:
Do you need to hash the node again here? You should be able to retrieve the hash from the node (I guess you need to probably rebase against master).

Reply:
I need the hash to check if it matches with any of the previous node. And I also need to get the rlp encoding to construct the recent witness.